### DISCUSSION TITLE
unable to hydrate my app

### DISCUSSION THREAD

**Discussion by leamandeep** [pmndrs/zustand#1877] [Q&A]
Hi Team! I'm unable to hydrate my app. I'm using `async storage` to persist my data. Need your guidance. Here's my snack link https://snack.expo.dev/@elamandeep/zustand?platform=web.

**Comment by leamandeep** [comment]
When i tried hydration in my local setup. My app crashed.

> Here's my zustand code 
```js
import create from "zustand";
import { persist, createJSONStorage } from "zustand/middleware";
import AsyncStorage from "@react-native-async-storage/async-storage";

export const useBearStore = create(
  persist(
    (set) => ({
      bears: 0,
      menuItems: [],
      user: {},
      category: [],
      catId: "",
      _hasHydrated: false,
      setHasHydrated: (state) => {
        set({
          _hasHydrated: state,
        });
      },
      setUser: (user) => set((state) => ({ user: user })),
      isLogin: false,
      setIsLogin: (isLogin) => set((state) => ({ isLogin: isLogin })),
      increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
      removeAllBears: () => set({ bears: 0 }),
      fontLoaded: false,
      changeFontLoaded: () => set((state) => ({ fontLoaded: true })),
      borderInputText: false,
      changeBorderInputText: () =>
        set((state) => ({ borderInputText: !state.borderInputText })),
      addMenuItems: (menuItems) => set((state) => ({ menuItems: menuItems })),
      addCategory: (category) => set((state) => ({ category: category })),
      setCatId: (catId) => set((state) => ({ catId: catId })),
    }),
    {
      name: "bear-storage",
      partialize: (state) =>
        Object.fromEntries(
          Object.entries(state).filter(([key]) =>
            ["user", "isLogin"].includes(key)
          )
        ),
      storage: createJSONStorage(() => AsyncStorage),
      onRehydrateStorage: () => (state) => {
        state.setHasHydrated(true);
      },
    }
  )
);
```

> App.js

```js
 const { isLogin, hasHydrated } = useBearStore((state) => ({
    isLogin: state.isLogin,
    hasHydrated: state._hasHydrated,
  }));
 
 console.log("hashydrated",hasHydrated)

  return !hasHydrated ? (
    <NhostProvider nhost={nhost}>
      <NhostApolloProvider nhost={nhost}>
        <SafeAreaView style={{ flex: 1 }}>
          <NavigationContainer>
            <PaperProvider>
              <ScreenStack.Navigator>
                {isLogin ? (
                  <ScreenStack.Screen
                    options={{ headerShown: false }}
                    name="App"
                    component={TabNavigator}
                  />
                ) : (
                  <>
                    <ScreenStack.Screen name="Login" component={Login} />
                    <ScreenStack.Screen name="Signup" component={Signup} />
                  </>
                )}
              </ScreenStack.Navigator>
            </PaperProvider>
          </NavigationContainer>
        </SafeAreaView>
      </NhostApolloProvider>
    </NhostProvider>
  ) : (
    <>
      <LoadingAnimation />
    </>
  );

```

![image](https://github.com/pmndrs/zustand/assets/68345524/8fe3c69e-9fd6-4113-ae59-13c579ba434e)

**Reply by dbritto-dev** [reply]
@elamandeep would you mind partializing just the user field? 

```ts
partialize: (state) => ({ user: state.user }),
```

I want to discard some possibilites

**Reply by leamandeep** [reply]
@dbritto-dev  Could you please double-check my snacks? It don't seem to be working as intended.

**Reply by dbritto-dev** [reply]
@elamandeep sure, I'll do it

**Reply by leamandeep** [reply]
@dbritto-dev please do it soon. I have to delivery app to my client. 

I want to know from u
`Is this issue is related to zustand?`

**Reply by dbritto-dev** [reply]
@elamandeep there's no issue with Zustand just a type from your side
![image](https://github.com/pmndrs/zustand/assets/1697714/345a7033-cb74-4933-ac30-5f9329a0c3ce)

**Reply by dbritto-dev** [reply]
You need to change `hashydrated` to `hasHydrated` and same for `_hashydrated` to `_hasHydrated` and `partialize: (state) => ({ login: state.isLogin }),` to `partialize: (state) => ({ login: state.login }),`

![image](https://github.com/pmndrs/zustand/assets/1697714/b57c5fff-6866-4616-98ff-39ecf78b8513)

**Reply by leamandeep** [reply]
> When i tried hydration in my local setup. My app crashed.
> 
> > Here's my zustand code 
> ```js
> import create from "zustand";
> import { persist, createJSONStorage } from "zustand/middleware";
> import AsyncStorage from "@react-native-async-storage/async-storage";
> 
> export const useBearStore = create(
>   persist(
>     (set) => ({
>       bears: 0,
>       menuItems: [],
>       user: {},
>       category: [],
>       catId: "",
>       _hasHydrated: false,
>       setHasHydrated: (state) => {
>         set({
>           _hasHydrated: state,
>         });
>       },
>       setUser: (user) => set((state) => ({ user: user })),
>       isLogin: false,
>       setIsLogin: (isLogin) => set((state) => ({ isLogin: isLogin })),
>       increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
>       removeAllBears: () => set({ bears: 0 }),
>       fontLoaded: false,
>       changeFontLoaded: () => set((state) => ({ fontLoaded: true })),
>       borderInputText: false,
>       changeBorderInputText: () =>
>         set((state) => ({ borderInputText: !state.borderInputText })),
>       addMenuItems: (menuItems) => set((state) => ({ menuItems: menuItems })),
>       addCategory: (category) => set((state) => ({ category: category })),
>       setCatId: (catId) => set((state) => ({ catId: catId })),
>     }),
>     {
>       name: "bear-storage",
>       partialize: (state) =>
>         Object.fromEntries(
>           Object.entries(state).filter(([key]) =>
>             ["user", "isLogin"].includes(key)
>           )
>         ),
>       storage: createJSONStorage(() => AsyncStorage),
>       onRehydrateStorage: () => (state) => {
>         state.setHasHydrated(true);
>       },
>     }
>   )
> );
> ```
> 
> > App.js
> 
> ```js
>  const { isLogin, hasHydrated } = useBearStore((state) => ({
>     isLogin: state.isLogin,
>     hasHydrated: state._hasHydrated,
>   }));
>  
>  console.log("hashydrated",hasHydrated)
> 
>   return !hasHydrated ? (
>     <NhostProvider nhost={nhost}>
>       <NhostApolloProvider nhost={nhost}>
>         <SafeAreaView style={{ flex: 1 }}>
>           <NavigationContainer>
>             <PaperProvider>
>               <ScreenStack.Navigator>
>                 {isLogin ? (
>                   <ScreenStack.Screen
>                     options={{ headerShown: false }}
>                     name="App"
>                     component={TabNavigator}
>                   />
>                 ) : (
>                   <>
>                     <ScreenStack.Screen name="Login" component={Login} />
>                     <ScreenStack.Screen name="Signup" component={Signup} />
>                   </>
>                 )}
>               </ScreenStack.Navigator>
>             </PaperProvider>
>           </NavigationContainer>
>         </SafeAreaView>
>       </NhostApolloProvider>
>     </NhostProvider>
>   ) : (
>     <>
>       <LoadingAnimation />
>     </>
>   );
> 
> ```
> 
> ![image](https://github.com/pmndrs/zustand/assets/68345524/8fe3c69e-9fd6-4113-ae59-13c579ba434e)
> 

@dbritto-dev  i did correctly. But why my app is crashing?

**Reply by dbritto-dev** [reply]
@elamandeep looks like something is causing a infinitve loop in your app, might be an `useEffect()` and that's why your app ended up crashing

**Reply by dbritto-dev** [reply]
@elamandeep BTW, you could try rehydrating your app manually by using `skipHydration: true` and using

```ts
useEffect(() => {
  useBearStore.persist.rehydrate(); 
}, [])
```

**Reply by leamandeep** [reply]
Thanks i will try

**Reply by dbritto-dev** [reply]
> Thanks i will try

@elamandeep keep me in the loop
